ccmlib/scylla_node.py

```diff
@@ -536,51 +537,77 @@ def start(self, join_ring=True, no_wait=False, verbose=False,
                                '%s.%s' % (socket.gethostname(), self.name)]
 
-        # Let's add jvm_args and the translated args
+        args = [launch_bin, '--options-file', options_file, '--log-to-stdout', '1']
 
-        args = [launch_bin, '--options-file', options_file, '--log-to-stdout', '1'] + jvm_args + translated_args
+        MB = 1024 * 1024
 
-        # Lets search for default overrides in SCYLLA_EXT_OPTS
-        scylla_ext_opts = os.getenv('SCYLLA_EXT_OPTS', "").split()
-        opts_i = 0
-        orig_args = list(args)
-        while opts_i < len(scylla_ext_opts):
-            if scylla_ext_opts[opts_i].startswith("--scylla-manager="):
-               opts_i += 1
-            elif scylla_ext_opts[opts_i].startswith('-'):
-                o = scylla_ext_opts[opts_i]
+        def process_opts(opts):
+            ext_args = OrderedDict()
+            opts_i = 0
+            while opts_i < len(opts):
+                # the command line options show up either like "--foo value-of-foo"
+                # or as a single option like --yes-i-insist
+                assert opts[opts_i].startswith('-')
+                if opts[opts_i].startswith("--scylla-manager="):
+                    opts_i += 1
+                    continue
+                o = opts[opts_i]
                 opts_i += 1
                 if '=' in o:
                     opt = o.replace('=', ' ', 1).split()
+                    key = opt[0]
+                    val = opt[1]
                 else:
-                    opt = [ o ]
-                    while opts_i < len(scylla_ext_opts) and not scylla_ext_opts[opts_i].startswith('-'):
-                        opt.append(scylla_ext_opts[opts_i])
+                    key = o
+                    vals = []
+                    while opts_i < len(opts) and not opts[opts_i].startswith('-'):
+                        vals.append(opts[opts_i])
                         opts_i += 1
-                if opt[0] not in orig_args:
-                    args.extend(opt)
+                    val = ' '.join(vals)
+                if key not in ext_args:
+                    ext_args[key] = val
+            return ext_args
+
+        # Lets search for default overrides in SCYLLA_EXT_OPTS
+        env_args = process_opts(os.getenv('SCYLLA_EXT_OPTS', "").split())
+
+        # precalculate self._mem_mb_per_cpu if --memory is given in SCYLLA_EXT_OPTS
+        # and it wasn't set explicitly by the test
+        if not self._mem_mb_set_during_test and '--memory' in env_args:
+            memory = self.parse_size(env_args['--memory'])
+            smp = int(env_args['--smp']) if '--smp' in env_args else self._smp
+            self._mem_mb_per_cpu = int((memory / smp) // MB)
+
+        cmd_args = process_opts(jvm_args)
+
+        # use '--memory' in jmv_args if mem_mb_per_cpu was not set by the test
+        if not self._mem_mb_set_during_test and '--memory' in cmd_args:
+            self._memory = self.parse_size(cmd_args['--memory'])
```

Comment:
this part is always the most time-consuming part when i read this file. i'd suggest refactor this part like:

```py
def get_option(self, name, default_value, env_args, cmd_args, customized):
    '''get an option from various source, in the order from lowest priority to
    highest, the option specified with higher priority override the one of lower
    priority'''
    if customized:
        return customized
    # settings specified by env variable have lowest priority, so check it first
    # settings specified by command line have higher priority, so check it later
    value = default_value
    for args in [env_args, cmd_args]:
        found = args.get(name)
        if found:
            try:
                value = int(found)
            except ValueError:
                value = self.parse_size(found)
    return value

smp = self.get_option('--smp',
                      self._smp,
                      env_args, cmd_args,
                      self._smp if self._smp_set_during_test else None)
default_memory = smp * self._mem_mb_per_cpu * MB
memory = self.get_option('--memory',
                         default_memory,
                         env_args, cmd_args,
                         default_memory if self._smp_set_during_test else None)
```

Reply:
It doesn't seem much simpler to me :-/
I think it might be over-engineered...
I wouldn't spend time on it unless there's something we'll need to fix  / improve around this code.